databricks/koalas/generic.py

```diff
@@ -2336,18 +2338,22 @@ def first_valid_index(self) -> Optional[Union[Scalar, Tuple[Scalar, ...]]]:
 
         cond = reduce(lambda x, y: x & y, map(lambda x: x.isNotNull(), data_spark_columns))
 
-        first_valid_row = (
-            self._internal.spark_frame.filter(cond)
-            .select(self._internal.index_spark_columns)
-            .first()
-        )
+        with sql_conf({SPARK_CONF_ARROW_ENABLED: False}):
+            # Disable Arrow to keep row ordering.
+            first_valid_row = (
+                self._internal.spark_frame.filter(cond)
+                .select(self._internal.index_spark_columns)
+                .limit(1)
```

Comment:
Can we mimic what we do in Koalas' DataFrame.head?

Reply:
Do you mean we should add `.orderBy(NATURAL_ORDER_COLUMN_NAME)` instead of disabling Arrow execution?

Reply:
I was just wondering ...

Reply:
Hm, I still don't get why the order gets changed though.

Reply:
The underlying plans are different between with and without Arrow as well as the execution method:

e.g.,

```py
sdf = spark.range(10)
sdf.limit(5).toPandas()
```

----

- Without Arrow

`collect()` -> `plan.executeCollect().map(fromRow)`

The plan is:

```
== Physical Plan ==
CollectLimit 5
+- *(1) Range (0, 10, step=1, splits=12)
```

There is no shuffle and `CollectLimitExec.executeCollect()` runs several Spark jobs **from the first partition** until the specified number of rows are collected. Thus the row order will be preserved.

----

- With Arrow:

`collectAsArrowToPython()` -> `toArrowBatchRdd(plan)` -> `plan.execute().mapPartitionsInternal ...`

The plan is:

```
== Physical Plan ==
*(2) Project [id#0L AS col_0#12L]
+- *(2) GlobalLimit 5
   +- Exchange SinglePartition, true, [id=#63]
      +- *(1) LocalLimit 5
         +- *(1) Range (0, 10, step=1, splits=12)
```

There is a shuffle to limit the number of rows which won't preserve the order of rows and `plan.execute()` will just return the generated RDD as-is.

Reply:
gotya. we gotta fix it in Apache Spark I guess .. anyway thanks. LGTM